PYIC-2080 Use CI scoring config to calculate user's overall CI score

blakeyp opened the pull request:
## Proposed changes

### What changed

Read the CI scoring values stored as json in a secret. For each of the user's stored CIs add up the "detected" scores to give an overall score. If this is greater than the allowed threshold stop the journey.

Also no longer check for the presence of CIs when recording a VC as "successful" or not - the profile score is sufficient and we don't want to fail the journey in the case that CIs are present but within the allowed threshold. The only exception is we still check for the presence of CIs when we send the "success" audit event on retrieving a credential from a CRI.

### Why did it change

Previously we were simply checking if there were any stored CIs and failing the journey on that basis (except for only-A01). Now we take into account the "detected" scores and only fail the journey if the total score exceeds the configured threshold (which won't be the case for only-A01). Eventually we will also handle the "checked" scores for doing mitigations. 

### Issue tracking
- [PYIC-2080](https://govukverify.atlassian.net/browse/PYIC-2080)

blakeyp committed:
PYIC-2080 Use CI scoring config to calculate user's overall CI score

Read the CI scoring values stored as json in a secret. For each of the user's stored
CIs add up the "detected" scores to give an overall score. If this is greater than
the allowed threshold stop the journey.

blakeyp committed:
PYIC-2080 Update config

blakeyp committed:
PYIC-2080 Update isVcSuccessful check to take into account CI scores, not just presence of CIs

dannyhayes-gds commented on lib/src/main/java/uk/gov/di/ipv/core/library/helpers/VcHelper.java:
```diff
@@ -46,23 +53,37 @@ public static boolean isSuccessfulVc(
             for (CredentialEvidenceItem item : credentialEvidenceList) {
                 if (item.getType().equals(CredentialEvidenceItem.EvidenceType.EVIDENCE)) {
-                    return Gpg45EvidenceValidator.isSuccessful(item);
+                    return Gpg45EvidenceValidator.isSuccessful(item, ciScoresMap, ciScoreThreshold);
```
I think for all these successful checks we can remove any checking of CI's. I don't think it really determines if a VC is successfull anymore, as if the CI is mitigated then it is successful.

I think these "success" checks should just be based on the actual gpg45 score value, i.e if it is 0 or not.

kerrr replied:
I think the reality is that apart from Fraud, VCs won't have a non zero profile score if they contain a CI. Lets just leave this for now.

dannyhayes-gds commented on lib/src/main/java/uk/gov/di/ipv/core/library/domain/gpg45/validation/Gpg45DcmawValidator.java:
```diff
@@ -1,21 +1,31 @@
+            return false;
+        }
+        if (item.getValidityScore() != 0) {
+            if (item.getCi() == null || item.getCi().isEmpty()) {
```
I think we can get rid of the ci scoring checks here now.

Should double check with @kerrr how we want to handle the "success" status of the audit events. So when we retrieve a new VC, we sent an audit event that contains that VC's success status that uses this code. For the audit message is it fine to exclude CI checks in that status as well, or for the audit message do we want to mark it as not successful if it contains a CI?

kerrr replied:
🤔 We shouldn't do CI scoring here. Even if we think that having a CI or not has some basing on success it's not the same issue as the scoring of CIs. As such it doesn't belong in this PR. So just leave this change out.

dannyhayes-gds replied:
I think a change might still need to be made potentially. This code is being called to trigger when a VC is considered successful or not during the select-cri lambda. Which probably doesn't need to count CI's into that fact anymore. We previously had it return false if a CI was present unless it was the Fraud CRI specifically with A01

blakeyp replied:
I've updated this now so we only check for presence of CIs for the VC success check in the retrieve-cri-credential lambda which uses it for the audit log. When we record VC success in the select-cri lambda we now ignore the CI check so the journey can continue.

dannyhayes-gds commented on lib/src/main/java/uk/gov/di/ipv/core/library/domain/gpg45/validation/Gpg45EvidenceValidator.java:
```diff
@@ -1,17 +1,28 @@
+            Map<String, ContraIndicatorScores> ciScoresMap,
+            int ciScoreThreshold) {
+        if (item.getValidityScore() != 0) {
+            if (item.getCi() == null || item.getCi().isEmpty()) {
```
Same here

dannyhayes-gds commented on lib/src/main/java/uk/gov/di/ipv/core/library/domain/gpg45/validation/Gpg45FraudValidator.java:
```diff
@@ -11,11 +15,16 @@ private Gpg45FraudValidator() {
+            int ciScoreThreshold) {
+        if (item.getIdentityFraudScore() != 0) {
+            if (item.getCi() == null || item.getCi().isEmpty()) {
+                return true;
```
Same here

dannyhayes-gds commented on lib/src/main/java/uk/gov/di/ipv/core/library/domain/gpg45/validation/Gpg45VerificationValidator.java:
```diff
@@ -1,17 +1,28 @@
+            Map<String, ContraIndicatorScores> ciScoresMap,
+            int ciScoreThreshold) {
+        if (item.getVerificationScore() != 0) {
+            if (item.getCi() == null || item.getCi().isEmpty()) {
```
Same here

dannyhayes-gds commented on lambdas/evaluate-gpg45-scores/src/main/java/uk/gov/di/ipv/core/evaluategpg45scores/EvaluateGpg45ScoresHandler.java:
```diff
@@ -210,7 +213,13 @@ private List<VcStatusDto> generateVcSuccessStatuses(List<SignedJWT> credentials)
+                    Integer.parseInt(configurationService.getSsmParameter(CI_SCORING_THRESHOLD));
+            boolean isSuccessful =
+                    VcHelper.isSuccessfulVc(
+                            signedJWT, addressCriConfig, ciScoresMap, ciScoreThreshold);
```
I don't think we should do the CI scoring here. It should be at the top level across all VC's gathered. Not against an individual VC

kerrr replied:
Yes, agreed, the scoring is not required here.

kerrr commented on lib/src/main/java/uk/gov/di/ipv/core/library/domain/gpg45/Gpg45ProfileEvaluator.java:
```diff
@@ -63,9 +65,20 @@ public Optional<JourneyResponse> getJourneyResponseForStoredCis(
+            ContraIndicatorScores scoresConfig = contraIndicatorScoresMap.get(ci);
+            ciScore += scoresConfig.getDetectedScore();
+        }
+        LOGGER.info("User's CI score: {}", ciScore);
```
We should always use structured logging with separate fields for values, rather than interpolating values into strings.

blakeyp replied:
Updated to use structured log

kerrr commented on lib/src/main/java/uk/gov/di/ipv/core/library/domain/ContraIndicatorScores.java:
```diff
@@ -0,0 +1,20 @@
+import lombok.Getter;
+
+@Getter
+public class ContraIndicatorScores {
```
Bike shedding, but this seems to represent a single score, so should be `ContraindicatorScore` singular.

blakeyp replied:
Updated

kerrr commented on lib/src/main/java/uk/gov/di/ipv/core/library/domain/ContraIndicatorScores.java:
```diff
@@ -0,0 +1,20 @@
+
+@Getter
+public class ContraIndicatorScores {
+    private String ci;
```
I'd be tempted to call this ciCode.

blakeyp replied:
I've left this as this is mapping the JSON object we're using in the config which uses `ci` as the field

blakeyp committed:
Revert "PYIC-2080 Update isVcSuccessful check to take into account CI scores, not just presence of CIs"

This reverts commit 8102668ae149a8d7226f4c9297f08220d825af7e.

blakeyp committed:
PYIC-2080 Address code review comments

Only include CIs in VC success check for the audit log in RetrieveCriCredentialHandler
Tidy up/rename

dannyhayes-gds commented on deploy/template.yaml:
```diff
@@ -1483,6 +1483,7 @@ Resources:
           IPV_SESSIONS_TABLE_NAME: !Ref SessionsTable
           USER_ISSUED_CREDENTIALS_TABLE_NAME: !Ref UserIssuedCredentialsV2Table
           CREDENTIAL_ISSUERS_CONFIG_PARAM_PREFIX: !Sub "/${Environment}/core/credentialIssuers"
+          CI_SCORING_THRESHOLD: !Sub "/${Environment}/core/self/ciScoringThreshold"
```
Do we need this ENV var? I think you've got it in the ConfigurationVariable class as well so I think that's the only one we need

blakeyp replied:
Good spot I've removed

dannyhayes-gds commented on lib/src/main/java/uk/gov/di/ipv/core/library/domain/gpg45/domain/CredentialEvidenceItem.java:
```diff
@@ -77,7 +77,7 @@ public Gpg45Scores.Evidence getEvidenceScore() {
     }
 
-    public boolean hasContraIndicators() {
+    public boolean hasContraIndicatorsExcludingFraudA01() {
```
I think this should be refactored slightly now. We no longer really need to be hard coding the fraud A01 scenario anymore.

I think this method should go back to its original name, and the logic inside that does the check of it being fraud etc can be removed. This method should just return true if any item has any CI at all

blakeyp replied:
So this is used for the VC success check for the audit event in retrieve-cri-credential - I was thinking it should still be considered "success" if it's fraud with only the A01 CI?

dannyhayes-gds replied:
I think that would be a change in the behaviour from what we currently have. Currently for the audit event, if it's fraud with A01 that is marked as not successful. Its only marked as successful for the select-cri bit

blakeyp replied:
Ah I see what you mean now yeah, I forgot about the flag being set to false there. Will update this then

dannyhayes-gds commented on lib/src/main/java/uk/gov/di/ipv/core/library/service/ConfigurationService.java:
```diff
@@ -230,6 +232,30 @@ public String getCriPrivateApiKey(String criId) {
+                GetSecretValueRequest.builder()
+                        .secretId(
+                                String.format(
+                                        "/%s/core/self/ci-scoring-config",
```
We should maybe move this string secrets manager var name into the ConfigurationVariable class

blakeyp replied:
Done

blakeyp committed:
PYIC-2080 Code review

blakeyp committed:
PYIC-2080 Fix code smells

blakeyp committed:
PYIC-2080 Code review - preserve previous behaviour, audit of any VC with CI is considered not successful
